keps/sig-network/20191104-iptables-no-cluster-cidr.md

```diff
@@ -0,0 +1,282 @@
+// accepted by the "kubernetes forwarding rules" rule above will be
+// accepted, to be as specific as possible the traffic must be sourced
+// or destined to the clusterCIDR (to/from a pod).
+writeLine(proxier.filterRules,
+	"-A", string(KubeForwardChain),
+	"-s", proxier.clusterCIDR,
+	"-m", "comment", "--comment", `"kubernetes forwarding conntrack pod source rule"`,
+	"-m", "conntrack",
+	"--ctstate", "RELATED,ESTABLISHED",
+	"-j", "ACCEPT",
+)
+writeLine(proxier.filterRules,
+	"-A", string(KubeForwardChain),
+	"-m", "comment", "--comment", `"kubernetes forwarding conntrack pod destination rule"`,
+	"-d", proxier.clusterCIDR,
+	"-m", "conntrack",
+	"--ctstate", "RELATED,ESTABLISHED",
+	"-j", "ACCEPT",
+)
+```
+Again, applying similar logic to the last rule for iptables, the proposal here is to simplify this drop reference to the proxy.clusterCIDR and just match on the connection state.
+
+### Risks and Mitigations
+
+The biggest risk we have is that we are expanding the scope of the last rule to potentially include non-kubernetes traffic. This is considered mostly safe as it does not break any of the intended drop behavior. Plus once the initial connection has been accepted, assuming nodes are used for kubernetes workloads, it's highly unlikely that we would need to not accept it later.
+
+## Design Details
+
+The idea of ‘determine cluster originated traffic’ would be captured in a new interface type within kube-proxy, with different implementations of the interface. The kube-proxy implementation itself would just call method on interface to get the match criteria to write in the rule.
+
+This assumes that the match can be represented in a single rule. We want to avoid going down the path, as much as possible, of adding multiple rules. Support for multiple rules will be taken as a separate enhancement based on feedback as the current clusterCIDR rule is not yet being removed.
+
+### Graduation Criteria
+
+TODO - this will be resolved once the initial KEP has been reviewed. The current thinking is that these would just be additional flags to kube-proxy.
+
+## Implementation History
+
+2019-11-04 - Creation of the KEP
+
+## Drawbacks [optional]
+
+The main caveat in this KEP is the relaxation of the accept rule for "ESTABLISHED,RELATED" packets. The other two rules have equivalent implementations, as long as we continue to guarantee that pod traffic is routed at the node boundary on _every_ and _all_ nodes that makes up the kubernetes cluster. This would not work if that assumption were to change.
+
+## Alternatives [optional]
+
+### Multiple cluster CIDR rules
+One alternative to consider is to explicitly track a list of cluster CIDRs in the ip table rules.
+For the second rule and third, it would be a simple source match to each of the CIDRs.
+But for the first rule, we would have to do a new mark and then masquerade on absence of mark, as we have to make sure it does not match _any_ of the allocated CIDRs.
```

Comment:
You don't need an extra mark:

    -A WHEREVER -blah -blah -blah -j MARK-MASQ-IF-NOT-LOCAL

    -A MARK-MASQ-IF-NOT-LOCAL -s 10.0.1.0/24 -j RETURN
    -A MARK-MASQ-IF-NOT-LOCAL -s 10.0.3.0/24 -j RETURN
    -A MARK-MASQ-IF-NOT-LOCAL -s 10.0.5.0/24 -j RETURN
    -A MARK-MASQ-IF-NOT-LOCAL -j KUBE-MARK-MASQ

(The same thing would work with multiple interface prefixes, etc, too.)

Reply:
This rule happens in the main chain for KUBE-SERVICE. So I don't think we can just RETURN here unless we move this to it's own CHAIN and jump to it. Trying to avoid creating new chains as much as possible.

Reply:
> Trying to avoid creating new chains as much as possible.

New chains are practically free. Mark bits are a limited resource.

Reply:
thanks for this. updated in the design details section.

Reply:
Having updated in implementation section, marking as resolved.